### ISSUE TITLE
Problem with createAddress with publicSpendKey of walletd

### ISSUE THREAD

**Issue by pmitchev** [amjuarez/bytecoin#71]
In walletd, when i create an address with publicSpendKey, I am able to spend it's funds with the first created address of the container

**Comment by pmitchev** [comment]
I found the problem. The documentation in the wiki is wrong. 
The correct parameter is "spendPublicKey"   NOT   "publicSpendKey".
I hope this will save time to others

**Comment by pmitchev** [comment]
Got it working. The process for using **publicSpendKey** in walletd is not well documented, so here it is:

1. Create the container as usual (with -g flag)
2. reset the wallet with **viewSecretKey** parameter
3. Create your wallets with publicSpendKey. 
4. Have a happy life

**Comment by KittyCatTech** [comment]
As mentioned above, the inputs are wrong in the documentation for "createAddress method:
"secretSpendKey" should be "spendSecretKey"
"publicSpendKey" should be "spendPublicKey"

Also,
1. Wallets can't be reset(to view only mode) with the same "viewSecretKey" that is currently in use. That's why you have to generate a new container.
2. 'View only' wallet containers (ones that have been reset with with "viewSecretKey") can not create (spend) addresses with "spendSecretKey", they can only create view only wallet address.
3. Likewise wallet containers with (spend) Addresses, can not create view only addresses.
4. The error messages when trying to create 'view only' addresses in 'spend' containers or trying to create 'spend' addresses on 'view only' containers are incorrect.

It would be nice to be able to a have both spend addresses and view addresses in the same container. Additionally or alternatively it would be nice to be able to input an address's "spendSecretKey" to the sendTransaction method to send from a 'view only' address or an address not stored in the container.

**Comment by pmitchev** [comment]
Wallets can not reset indeed. I am currently working on a fix before the Bytecoin team make another release.
If you want to test the view keys functionality you can use v2.0.0 version. Otherwise you have to wait for a fix

**Comment by KittyCatTech** [comment]
I should have mentioned I am using Forknote 2.0.1. That is the explanation of what it is doing.

**Comment by pmitchev** [comment]
Problem found, walletd just freezes on exit/reset when you it's not connected to a daemon.

**Comment by KittyCatTech** [comment]
I don't think you understood my post, I am able to make view only addresses, but I can't also create spend addresses in the same container.

**Comment by pmitchev** [comment]
In 1 container you can get only 'view addresses' or 'spend addresses', not both

**Comment by KittyCatTech** [comment]
Why not both?

**Comment by aivve** [comment]
For security reasons. View only container is view only.

**Comment by indigencoin** [comment]
Hi, @KittyCatTech how do you exactly specify the spendSecretKey in a sendTransaction when that parameter is not being asked by sendTransaction method?

**Comment by indigencoin** [comment]
Let me rephrase my question, how exactly do you use a spendSecretKey with sendTransaction to send funds? I still can't figure out why I can't spend the funds from an address contained in a wallet container. I always get "Invalid Request" error code: -32600

**Comment by KittyCatTech** [comment]
To use the spendSecretKey you call the 'createAddress' method with the 'spendSecretKey' parameter. This adds that address to the container. Once the address is in the container, then you use the 'sendTransaction' method with the 'addresses' parameter which needs to be an array of addresses that you are sending from.

**Comment by indigencoin** [comment]
walletservice.cpp (line 874) is now using sourceAddresses parameter. So I changed it to that but I still get the same error. I tried to locate where this error -32600 is coming from and from the file PaymentServiceJsonRpcServer.h I found this:

`      try {
        CryptoNote::JsonInputValueSerializer inputSerializer(const_cast<Common::JsonValue&>(jsonRpcParams));
        serialize(request, inputSerializer);
      } catch (std::exception&) {
        makeGenericErrorReponse(jsonResponse, "Invalid Request", -32600);
        return;
      }`

The serializer keeps throwing an exception even though my json is properly formatted:

{
  "jsonrpc": "2.0",
  "id": "",
  "method": "sendTransaction",
  "params": {
    "sourceAddresses": [ "indgn7KayssFJVj4bhzv6LbdG8RF5h3FJJ6utXGqpwAQijBg6GRkCxELHfpZKdNFr8fEF9cPizJGea3YvGsVzppm9NKT8X7vNBT38" ],
    "transfers": [
      {
        "amount": 70000000,
        "address": "indgn7KJrSR93gvYFocGqKDiWZRdsV4XwjMzTuk6nfyERKe2noXdWyVWCFBX4XswkghwvxbZqGKPKBPhNWSG841E7bbfb1rGdUX25"
      }
    ],
    "fee": 10000,
    "unlockTime": 1,
    "anonymity": 7,
    "extra": "",
    "paymentId": "24",
    "changeAddress": ""
  },
  "result": { "transactionHash": null }
}

**Comment by pmitchev** [comment]
Your paymentId is wrong. Try to leave it blank to check is that the only error

**Comment by indigencoin** [comment]
Same error.  passed blank value to paymentId, removed result class. Been experimenting with the json file, I just keep getting error -32600

**Comment by indigencoin** [comment]
BTW, I created the wallet address with both blank spendSecretKey and spendPublicKey. I never execute the reset() the wallet with viewSecretKey. So I assume this is has spendable wallets and not view only wallets.

**Comment by KittyCatTech** [comment]
https://wiki.bytecoin.org/wiki/Send_transaction_-_Bytecoin_RPC_Wallet_API

**Comment by indigencoin** [comment]
I tried both addresses and sourceAddresses, nothing's working. I tried resetting the container, and I was able to create 4 addresses with public and secret spend keys. But still get same error. So there must be one parameter that is causing the error. But looking at these codes, there are some inconsistencies:

** this one uses sourceAddresses
https://github.com/amjuarez/bytecoin/blob/master/src/PaymentGate/PaymentServiceJsonRpcMessages.h#L305

** this one still uses addresses
https://github.com/amjuarez/bytecoin/blob/master/src/PaymentGate/PaymentServiceJsonRpcMessages.cpp#L230

**Comment by indigencoin** [comment]
Found a fix. It was paymentId parameter that is wrong. It should be payment_id.

But it's kind of strange because both addresses and sourceAddresses are working and valid so you can use either of the two. However, if you use sourceAddresses, it requires changeAddress parameter, while if you use addresses only, it is not required.

**Comment by pmitchev** [comment]
after I read one of your code containing comments - "addresses" is the right one
